packages/verkle/src/verkleTree.ts

```diff
@@ -213,11 +215,168 @@ export class VerkleTree {
    * @param value - the value to store
    * @returns A Promise that resolves once value is stored.
    */
-  // TODO: Rewrite following logic in verkle.spec.ts "findPath validation" test
-  async put(_key: Uint8Array, _value: Uint8Array): Promise<void> {
-    throw new Error('not implemented')
+  async put(key: Uint8Array, value: Uint8Array): Promise<void> {
+    if (key.length !== 32) throw new Error(`expected key with length 32; got ${key.length}`)
+    const stem = key.slice(0, 31)
+    const suffix = key[key.length - 1]
+    this.DEBUG && this.debug(`Stem: ${bytesToHex(stem)}; Suffix: ${suffix}`, ['PUT'])
+
+    const putStack: [Uint8Array, VerkleNode][] = []
+    // Find path to nearest node
+    const foundPath = await this.findPath(stem)
+
+    // Sanity check - we should at least get the root node back
+    if (foundPath.stack.length === 0) {
+      throw new Error(`Root node not found in trie`)
+    }
+
+    // Step 1) Create or update the leaf node
+    let leafNode: LeafNode
+    // First see if leaf node already exists
+    if (foundPath.node !== null) {
+      // Sanity check to verify we have the right node type
+      if (!isLeafNode(foundPath.node)) {
+        throw new Error(
+          `expected leaf node found at ${bytesToHex(stem)}. Got internal node instead`
+        )
+      }
+      leafNode = foundPath.node
+      // Sanity check to verify we have the right leaf node
+      if (!equalsBytes(leafNode.stem, stem)) {
+        throw new Error(
+          `invalid leaf node found. Expected stem: ${bytesToHex(stem)}; got ${bytesToHex(
+            foundPath.node.stem
+          )}`
+        )
+      }
+    } else {
+      // Leaf node doesn't exist, create a new one
+      leafNode = await LeafNode.create(
+        stem,
+        new Array(256).fill(new Uint8Array(32)),
+        this.verkleCrypto
+      )
+      this.DEBUG && this.debug(`Creating new leaf node at stem: ${bytesToHex(stem)}`, ['PUT'])
+    }
+    // Update value in leaf node and push to putStack
+    leafNode.setValue(suffix, value)
+    this.DEBUG &&
+      this.debug(
+        `Updating value for suffix: ${suffix} at leaf node with stem: ${bytesToHex(stem)}`,
+        ['PUT']
+      )
+    putStack.push([leafNode.hash(), leafNode])
+
+    // `path` is the path to the last node pushed to the `putStack`
+    let lastPath = leafNode.stem
+
+    // Step 2) Determine if a new internal node is needed
+    if (foundPath.stack.length > 1) {
```

Comment:
Just to make sure I am getting this correctly, could you confirm that this scenario is correct:
- Let's say we start with a root node
- We put a key/value pair where the key starts with `0x01`. 
- `findPath` returns the rootNode.
- We create a leafNode, and add a pointer to the newly created leaf node at index `2` of the root node. 
- We put another key/value pair that also starts with `0x01`. 
- `findPath` now returns the `leafNode` that was created in the previous put operation, along with the `remaining` values where the paths branch off. 
- We then create an internal node where the paths branch off, commit to the two children to it, and then commit the internal node to the root node (at index `2`).
- rootHash gets updated.
-

Reply:
That should be how it works. Not guaranteeing I got it exactly right but the tests seem to indicate so

Reply:
Cool, I didn't infer the above from how the function/tests were structured, but wrote it based on my understanding of verkle trees, good to hear it matches

Reply:
Looking at your comment again, let me confirm something.  What you call "index 2" here would be `rootNode.children[1]` right?  These are zero indexed arrays and so inserting a node that starts with `0x01` means you check `rootNode.children[1]` to see if there's an existing reference to some child node and if not, create a new leaf node with the value for the key that starts with `0x01` and then put the commitment from that new leaf node in `rootNode.children[1]`

Reply:
Ah yes that's what I meant, sorry for the confusing indexing